### PR TITLE
Identify overloaded symbols

### PR DESCRIPTION
**Pull request by mayaepps**
Bug/issue #, if applicable: rdar://117092401

## Summary

This is a PR to improve the presentation of overloaded symbols in Swift DocC related to this [forum post](https://forums.swift.org/t/improving-the-presentation-of-overloaded-symbols-in-swift-docc/67713). Specifically, this PR will:
1. identify overloaded symbols,
2. encode each individual overloaded symbol with the necessary information to display its sibling overloads

## Testing

Steps:
1. Build documentation for a target that has overloaded symbols. 
2. Open the .doccarchive and navigate to the Render JSON for one of the overloaded symbols. 
3. Check that the JSON contains an `otherDeclarations` array where each element contains an `identifier` and an array of `tokens`.

## Checklist

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [ ] Updated documentation if necessary

### CODE REVIEW

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 1]
```diff
@@ -2350,6 +2354,100 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+                    
+                    (documentationNode.semantic as! Symbol).otherDeclarationsVariants = .init(swiftVariant: otherOverloadDeclarations)
+                }
+            }
+        }
+    }
+    
+    /// Returns a name for an overload.
+    /// If two symbols have the same name, they are overloads of eachother.
+    private func getOverloadUnqiueName(for node: DocumentationNode) -> String? {
```
Comment (ethan-kusters) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2439]: Nit: Generally Swift API naming avoids the `get` prefix. Maybe something like this?

```suggestion
    private func uniqueName(forOverloadedSymbol node: DocumentationNode) -> String? {
```

The Swift API design guidelines are super helpful for this kind of thing: https://www.swift.org/documentation/api-design-guidelines/.

File: Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift [hunk 2]
```diff
@@ -28,39 +29,55 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
...
                     guard let lhsValue = lhs, let rhsValue = rhs else {
                         return lhs == nil
                     }
                     return lhsValue.rawValue < rhsValue.rawValue
                 }
                 
+                let renderedTokens = renderDeclarationTokens(fragments: declaration.declarationFragments, renderNodeTranslator: &renderNodeTranslator)
+                
+                // If this symbol has overloads, render their declarations as well.
+                // TODO: Handle objc overloads?
```
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift:42]: Objective-C doesn't have name overloads because two methods with the same name also have the same selector. See [the archived The Objective-C Programming Language][1] if you're curious.

[1]: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/Introduction/Introduction.html#//apple_ref/doc/uid/TP40011210-CH1-SW1

That said, it'd still be good to process all the variants here to support other source languages in the future.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 3]
```diff
@@ -2350,6 +2354,100 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+    private func handleOverloads(relationships: [SymbolGraph.Relationship]) {
+        guard FeatureFlags.current.isExperimentalOverloadedSymbolPresentationEnabled else {
+            return
+        }
+        
+        var parentToChildren = [SymbolGraph.Symbol.Identifier : [DocumentationNode]]()
+        
+        // Walk the symbol graph relationships and look for parent <-> child links.
+        // Build up the parentToChildren dictionary of all children under each parent.
+        for relationship in relationships {
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2368]: It's also worth noting that not all symbol relationships contribute to the link / name hierarchy directly. For example, a "default implementation of" relationship is between the default implementation and the protocol requirement. For example

```swift
protocol SomeProtocol {
  func something() // the "protocol requirement"
}

extension SomeProtocol {
  func something() { // the "default implementation"
    // do something here
  } 
}
```

Adding the default implementation as a child of the protocol requirement would give it a path of "SomeProtocol/something()/something()" but we'd expect both symbols to have a path of "SomeProtocol/something()" which means that the default implementation symbol should be added as a child of the protocol requirement's parent symbol. 

There are a few other relationship kinds that follow this pattern which is all the more reason to avoid building up another symbol hierarchy based on relationships.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 4]
```diff
@@ -2350,6 +2354,100 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+        
+        // Types of symbols that we will collapse overloads for.
+        let overloadableTypes: [DocumentationNode.Kind] = [.instanceMethod, .typeMethod, .function,
+                                                           .macro, .operator, .initializer, .typeSubscript,
+                                                           .instanceSubscript]
+        guard overloadableTypes.contains(node.kind) else {
+            return nil
+        }
+        
+        return "\(node.kind.id)-\(node.name)"
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2449]: This only really matters if this name is ever displayed to the user somehow (for example in a URL component). If it's only used to group symbols by their names and kinds, then a private struct would be more descriptive.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2368]: That's a good point. The link resolver sounds like it would be great to use if it already knows the symbol collisions. Is there a way of using the link resolver to find the symbol collisions before links are actually resolved? I was hoping to recognize overloads and add a new node to the topic graph that represents a group of overloads before links were resolved. That way when writers link to a merged group of overloads (using a link without a hash) the link resolver might be more easily taught to resolve the link because the corresponding node for that link would already exist. However, I'm less familiar with how link resolution works, so I might be making incorrect assumptions about the best way to help resolve overload links without hashes.

[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2368]: PS:
I didn't know this 2 years ago but I've l've learnt the hard way by also writing this code and then gradually fixing bugs in it.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2449]: It's used to group symbols by their names and kinds to ensure that overloads are always of the same kind. A private struct sounds like a better way to go.

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift [hunk 5]
```diff
@@ -163,22 +163,57 @@ public struct DeclarationRenderSection: Codable, Equatable {
         }
     }
     
+    /// A declaration for a different symbol that is connected to this one, e.g. an overload.
+    public struct OtherDeclaration: Codable, Equatable, Hashable {
+        /// The overloaded symbol's declaration tokens.
+        public let tokens: [Token]
+        
+        /// The overloaded symbol's identifier.
+        public let identifier: ResolvedTopicReference
```
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift:172]: It is intended to be used to reference another page. A renderer would use this to look up content unique to that particular overload.

[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift:172]: From a quick look, it appeared that Strings were all that were used in other cases (such as [Tokens](https://github.com/apple/swift-docc/blob/5ad35a3107ca0443b81ada917b73b950d89bf396/Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift#L113))--do you think a ResolvedTopicReference should be used here?

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2368]: The link resolver is built up incrementally as more and more content is registered with the documentation context. All the symbol information is added when the symbol graphs are loaded (before the symbols are registered with the context) so if you only need to know about symbols then you can rely on all the symbols and symbol collisions being known.

The core of the link resolution implementation is a tree-like data structure—that's aware of collisions and disambiguation—which is controlled by `PathHierarchy`. Nothing interacts with it directly except for `PathHierarchyBasedLinkResolver` which provides an API to the rest of DocC and translates between `ResolvedTopicReference` values and the path hierarchy's internal identifiers. 

The "link resolver" is mainly responsible for finding what page a given link refers to (PathHierarchy+Find.swift) but since it knowns the most about collisions and disambiguation it also determines the paths of all the symbol pages (PathHierarchy+DisambiguatedPaths.swift).

To separate responsibilities, the "link resolver" (referring to `PathHierarchyBasedLinkResolver` and everything it wraps) doesn't call out to the rest of DocC, for example update the topic graph. Instead it provides API that provide the data that other parts of DocC need to make those modifications. For example: `traverseSymbolAndParentPairs(_:)` is used to do auto-curation.

If you want to perform some action on all the symbols that have symbol kind collisions you could possibly add something like this to `PathHierarchyBasedLinkResolver `:

```swift
func traverseSymbolAndSomething(_ observe: (_ symbol: ResolvedTopicReference, _ something: [ResolvedTopicReference]) -> Void) {
    for (id, node) in pathHierarchy.lookup {
        guard node.symbol != nil, let reference = resolvedReferenceMap[id] else { continue }
        
        // This is peeking into implementation details of how the path hierarchy internally stores disambiguation.
        // Once we know what data and what structure we need we should probably create a PathHierarchy API that hides those implementation details.
        for (_ /*name*/, disambiguation) in node.children {
            for (_ /*kind*/, innerStorage) in disambiguation.storage where innerStorage.count > 1 {
                let childrenWithSameKind = innerStorage.values.compactMap {
                    // Filter out any non-symbols and pages that don't have a known reference
                    $0.symbol == nil ? nil : resolvedReferenceMap[$0.identifier]
                }
                // ???: Does this need the parent reference or just the children? Does it need the name and/or kind?
                observe(reference, childrenWithSameKind)
            }
        }
    }
}
```

That said, at some point the core implementation will need to have some awareness of these combined overload pages so that that can be taken into account when resolving links. 

The code above is probably a decent starting point but I would expect it to change once the core implementation knowns more about combined overload pages.

File: Tests/SwiftDocCTests/Rendering/DeclarationsRenderSectionTests.swift [hunk 6]
```diff
@@ -52,7 +63,8 @@ class DeclarationsRenderSectionTests: XCTestCase {
                     DeclarationRenderSection(
                         languages: nil,
                         platforms: [],
-                        tokens: [.init(text: "", kind: token)]
+                        tokens: [.init(text: "", kind: token)],
+                        otherDeclarations: [DeclarationRenderSection.OtherDeclaration(tokens: [.init(text: "", kind: token)], identifier: "identifier")]
```
Comment (ethan-kusters) [inline Tests/SwiftDocCTests/Rendering/DeclarationsRenderSectionTests.swift:67]: Maybe we should also have a test that shows that we _don't_ emit an `otherDeclarations` section if this is empty?

**Review by ethan-kusters — COMMENTED** [review]
Looks really great. Just left a few nit-picks and comments.

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift [hunk 7]
```diff
@@ -163,22 +163,57 @@ public struct DeclarationRenderSection: Codable, Equatable {
...
+        
+        public func encode(to encoder: Encoder) throws {
```
[the remark this answers is not in the document]
Comment (ethan-kusters) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift]: Oh sorry – seeing now that you're only encoding the `absoluteString`. This will probably break round-trip encoding, is that expected?

**Comment by ethan-kusters** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift [hunk 8]
```diff
@@ -58,6 +58,26 @@ final class PathHierarchyBasedLinkResolver {
...
     
+    /// Traverse all symbols of the same kind that have collisions.
+    func traverseOverloadedSymbols(_ observe: (_ overloadedSymbols: [ResolvedTopicReference]) -> Void) {
+        let overloadableKinds = ["method", "type.method", "func", "init", "macro", "subscript", "func.op"]
+        
+        for (_, node) in pathHierarchy.lookup {
+            guard node.symbol != nil else { continue }
+            
+            // TODO: Once we know what data and what structure we need we should probably create a PathHierarchy API that hides those implementation details.
+            for (_ /*name*/, disambiguation) in node.children {
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift:69]: That's my fault. I suggested some code and didn't know if the name and kind values would be needed or not and I didn't want to introduce warnings about those variables never being used. If we don't need the name here we can remove the comment.

```suggestion
            for (_, disambiguation) in node.children {
```

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift]: This is actually outdated at this point--I switched it to be a `String`, and it shouldn't break round-trip encoding now. However, now I think the custom encoder is still needed to only encode `otherDeclarations` if it's not empty (unless the synthesized one is smart enough to do that automatically?)

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift:69]: I originally left it because I found it helpful to understand what was being iterated through 😄. It's not used though, so I'll remove it.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 9]
```diff
@@ -2350,6 +2352,31 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
     }
+
+    /// Handles overloaded symbols by grouping them together into one page.
+    private func handleOverloads(with linkResolver: PathHierarchyBasedLinkResolver) {
+        guard FeatureFlags.current.isExperimentalOverloadedSymbolPresentationEnabled else {
+            return
+        }
+        
+        linkResolver.traverseOverloadedSymbols { overloadedSymbolReferences in
+            guard overloadedSymbolReferences.count > 1 else { return }
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2363]: Maybe instead of guarding against "groups" containing zero or one overloaded symbols, we could update the logic in `traverseOverloadedSymbols` to guarantee that it only calls the closure for groups with 2 or more elements. In other words, maybe the concept of "overloaded symbols" doesn't make sense when there are none or just one.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 10]
```diff
@@ -2350,6 +2352,31 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+            for (index, symbolReference) in overloadedSymbolReferences.indexed() {
+                let documentationNode = try? entity(with: symbolReference)
+                
+                assert(documentationNode?.semantic is Symbol,
+                       "Overloads must be symbols.")
+                guard let symbol = (documentationNode?.semantic as? Symbol) else { continue }
+                
+                var otherOverloadSymbols = overloadedSymbolReferences
+                otherOverloadSymbols.remove(at: index)
+                symbol.overloadsVariants = .init(swiftVariant: otherOverloadSymbols)
```
[the remark this answers is not in the document]
Comment (patshaughnessy) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2375]: Or maybe we should add another `assert`? We would want this to fail/crash for debug builds at least.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift [hunk 11]
```diff
@@ -58,6 +58,26 @@ final class PathHierarchyBasedLinkResolver {
         }
     }
     
+    /// Traverse all symbols of the same kind that have collisions.
+    func traverseOverloadedSymbols(_ observe: (_ overloadedSymbols: [ResolvedTopicReference]) -> Void) {
+        let overloadableKinds = ["method", "type.method", "func", "init", "macro", "subscript", "func.op"]
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift:63]: Is there a better place or way to define this list? For example in 'SymbolGraph.Symbol.KinIdentifier'? But I recall @bitjammer talking about how the symbol graph format should not be tied to a specified programming language so maybe somewhere in DocC's source code would be better?

Anyway this list seems important enough it should be public and defined in some more obvious place.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift (hunk 8 above)
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift:69]: I wonder if these nested loops should be a method on `PathHierarchy.Node` that calls a closure. This code knows a lot about the internal structure of these nodes and maybe should be encapsulated in that struct so we don't leak that structure.

File: Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift [hunk 12]
```diff
@@ -28,39 +29,61 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
...
                         return lhs == nil
                     }
                     return lhsValue.rawValue < rhsValue.rawValue
                 }
                 
+                let renderedTokens = renderDeclarationTokens(fragments: declaration.declarationFragments, 
+                                                             renderNodeTranslator: &renderNodeTranslator)
+
+                // If this symbol has overloads, render their declarations as well.
+                let overloadsVariant = symbol.overloadsVariants[trait]
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift:43]: Maybe extra the following lines of code into a separate method similar to `renderDeclarationTokens` ? Then you could simplify this method down substantially into a single call to `declarations.append(....`

File: Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift [hunk 13]
```diff
@@ -28,39 +29,61 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
...
+                // TODO: Support overloads that have platform variants
+                let otherDeclarations = overloadsVariant?.compactMap { overloadReference -> DeclarationRenderSection.OtherDeclaration? in
+                    guard let overload = try? renderNodeTranslator.context.entity(with: overloadReference).semantic as? Symbol,
+                          let declarationFragments = overload.declarationVariants[trait]?.values.first?.declarationFragments else {
+                        return nil
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift:48]: Should we assert here? What does it mean if we get an overload that is not a symbol here, or that does not have any declaration fragments? Is that a valid scenario? Are there any symbols with no declaration at all?

Silently skipping over invalid symbols might make bugs elsewhere in the code harder to find.

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift [hunk 14]
```diff
@@ -163,22 +163,61 @@ public struct DeclarationRenderSection: Codable, Equatable {
         }
     }
     
+    /// A declaration for a different symbol that is connected to this one, e.g. an overload.
+    public struct OtherDeclaration: Codable, Hashable {
```
[the remark this answers is not in the document]
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift:167]: Maybe this should be a Swift Question of the Day... but does the lack of an initializer prevent clients from creating instances of this struct (or any struct)? Doesn't Swift implicitly create a default initializer for structs? Or maybe it does but they are not public?

**Review by patshaughnessy — APPROVED** [review]
Looks great to me. I left some coding style questions and suggestions.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift (hunk 11 above)
Comment (mayaepps) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchyBasedLinkResolver.swift:63]: I was just thinking about this :D
Somewhere in DocC sounds right to me. I'm not sure the best place to put it though--maybe in the PathHierarchy, since that's currently where DocC determines symbol name collisions?

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift (hunk 14 above)
[the remark this answers is not in the document]
Comment (ethan-kusters) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift:167]: That's exactly it- Swift will synthesize an `internal` initializer for structs by default, but if you want a public one you have to define it yourself.

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Convert.swift [hunk 15]
```diff
@@ -537,6 +537,12 @@ extension Docc {
             )
             var enableExperimentalLinkHierarchySerialization = false
             
+            @Flag(
+                name: .customLong("enable-experimental-overloaded-symbol-presentation"),
+                help: ArgumentHelp("Collects all the symbols that are overloads of each other onto a new merged-symbol page.")
+            )
+            var enableExperimentalOverloadedSymbolPresentation = false
```
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Convert.swift:544]: Should this be added now, or would it be better to wait until the feature is completed?

File: Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift (hunk 13 above)
Comment (mayaepps) [reply Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift]: I have a check back in the DocumentationContext that each overload must be a symbol, do you think I should assert here as well? 
I've added an assertion that all overloads must have declaration fragments.

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift]: It seems like the renderer is able to reference other pages using just this string, so I am resolving this for now.

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Convert.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Convert.swift:544]: @d-ronnqvist Is adding to the features.json something that should happen now, or is it better to wait until the feature is completed and usable?

**Review by d-ronnqvist — CHANGES_REQUESTED** [review]
I left a handful of comments. The most important how the overloads / other declarations are modeled. Since that's both public API and part of the render node spec it's very difficult to change after the fact.

File: Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Convert.swift (hunk 15 above)
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocCUtilities/ArgumentParsing/Subcommands/Convert.swift:544]: Added to features.json in [619ef63](https://github.com/apple/swift-docc/pull/740/commits/619ef63240ff83417a69198ed27c18320ba177ee)

File: Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift [hunk 16]
```diff
@@ -163,22 +163,67 @@ public struct DeclarationRenderSection: Codable, Equatable {
...
+            self.tokens = tokens
+            self.identifier = identifier
+        }
+    }
+    
+    /// The declarations for this symbol's overloads.
+    public let otherDeclarations: [OtherDeclaration]
+    
+    /// If this symbol has overloads, this symbol's declaration should be displayed at this index among the other overloads on the page.
+    public let indexInOtherDeclarations: Int?
```
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Model/Rendering/Symbol/DeclarationsRenderSection.swift]: That makes sense and probably would have prevented a bug I ran into a little while ago. I changed the RenderJSON and spec to reflect this in [619ef63](https://github.com/apple/swift-docc/pull/740/commits/619ef63240ff83417a69198ed27c18320ba177ee)

File: Sources/SwiftDocC/Semantics/Symbol/Symbol.swift [hunk 17]
```diff
@@ -568,4 +578,16 @@ extension Symbol {
...
+    var overloads: [ResolvedTopicReference] {
+        get { overloadsVariants.firstValue! }
+        set { overloadsVariants.firstValue = newValue }
+    }
+    
+    /// If this symbol has overloads, this symbol's declaration should be displayed at this index among the other overloads on the page.
+    var indexInOverloads: Int {
+        get { indexInOverloadsVariants.firstValue! }
+        set { indexInOverloadsVariants.firstValue = newValue }
+    }
```
[the remark this answers is not in the document]
Comment (mayaepps) [reply Sources/SwiftDocC/Semantics/Symbol/Symbol.swift]: Removed in [619ef63](https://github.com/apple/swift-docc/pull/740/commits/619ef63240ff83417a69198ed27c18320ba177ee).

File: Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift [hunk 18]
```diff
@@ -3918,17 +3918,20 @@ let expected = """
                 let overloadedDocumentationNode = try XCTUnwrap(context.documentationCache[reference])
                 let overloadedSymbol = try XCTUnwrap(overloadedDocumentationNode.semantic as? Symbol)
                 
+                XCTAssertNotNil(overloadedSymbol.overloadsVariants.firstValue)
+                let overloads = overloadedSymbol.overloadsVariants.firstValue!
```
Comment (d-ronnqvist) [inline Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift:3922]:
```suggestion
                let overloads = try XCTUnwrap(overloadedSymbol.overloadsVariants.firstValue)
```

File: Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift [hunk 19]
```diff
@@ -79,10 +78,16 @@ struct DeclarationsSectionTranslator: RenderSectionTranslator {
...
     
     /// Translates the given resolved topic references into encodable `OtherDeclaration`s representing the declarations of this symbol's overloads.
-    func renderOtherDeclarationsTokens(from overloads: [ResolvedTopicReference], for trait: DocumentationDataVariantsTrait,
-                                       renderNodeTranslator: inout RenderNodeTranslator) -> [DeclarationRenderSection.OtherDeclaration] {
-        var otherDeclarations = [DeclarationRenderSection.OtherDeclaration]()
-        for overloadReference in overloads {
+    func renderOtherDeclarationsTokens(from overloads: Symbol.Overloads?,
+                                       for trait: DocumentationDataVariantsTrait,
+                                       renderNodeTranslator: inout RenderNodeTranslator) -> DeclarationRenderSection.OtherDeclarations? {
+        guard let overloads else {
```
Comment (d-ronnqvist) [inline Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift:84]: We're still using an older Swift version that doesn't support this language feature 

```suggestion
        guard let overloads = overloads else {
```

Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/Rendering/RenderSectionTranslator/DeclarationsSectionTranslator.swift:84]: We could also keep the parameter non-optional and make it the callers responsibility to unwrap the value. For example, the caller could use [Optional.map(_:)](https://developer.apple.com/documentation/swift/optional/map(_:)) to only pass the variant if there's a value and return `nil` otherwise

```swift
let otherDeclarations = symbol.overloadsVariants[trait].map { variant in 
    renderOtherDeclarationsTokens(from: variant, for: trait, renderNodeTranslator: &renderNodeTranslator)
}

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test

**Comment by QuietMisdreavus** [comment]
I've taken over this PR, at @mayaepps' request. @d-ronnqvist, can you take another look? I should have resolved all the remaining review comments.

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test

**Comment by d-ronnqvist** [comment]
Yes, I can look at this again early next week.

**Review by d-ronnqvist — APPROVED** [review]

File: Sources/SwiftDocC/Infrastructure/Extensions/KindIdentifier+Overloads.swift [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+
+extension SymbolGraph.Symbol.KindIdentifier {
+    /// The kinds of symbols whose documentation pages should be grouped as overloads.
+    static let overloadableKinds: Set<SymbolGraph.Symbol.KindIdentifier> = [
+        .method,
+        .typeMethod,
+        .`func`,
+        .`init`,
+        .macro,
+        .subscript,
```
[the remark this answers is not in the document]
Comment (QuietMisdreavus) [reply Sources/SwiftDocC/Infrastructure/Extensions/KindIdentifier+Overloads.swift:21]: This could be handled in a separate PR, to make sure we also add test data at the same time.

File: Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift [hunk 21]
```diff
@@ -4256,6 +4256,92 @@ let expected = """
         problem = try XCTUnwrap(linkResolutionProblems.last)
         XCTAssertEqual(problem.diagnostic.summary, "\'NonExistingDoc\' doesn\'t exist at \'/BestBook/MyArticle\'")
     }
+    
+    func testContextRecognizesOverloads() throws {
```
[the remark this answers is not in the document]
Comment (QuietMisdreavus) [reply Tests/SwiftDocCTests/Infrastructure/DocumentationContext/DocumentationContextTests.swift:4260]: I think there could probably be a larger effort to build "mock symbol graph" functionality and replace some of the test data with that. (It would also be cool to have a test harness like the Swift compiler's lit tests or the source-compatibility suite that was able to build the symbol graphs from Swift/etc code itself, but that's a larger infrastructural effort...) Regardless, that's probably better served as its own PR or series of PRs, rather than holding up this PR.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 22]
```diff
@@ -2398,6 +2400,37 @@ public class DocumentationContext: DocumentationContextDataProviderDelegate {
...
+            for (index, symbolReference) in overloadedSymbolReferences.indexed() {
+                let documentationNode = try entity(with: symbolReference)
+
+                guard let symbol = documentationNode.semantic as? Symbol else {
+                    preconditionFailure("""
+                    Only symbols can be overloads. Found non-symbol overload for \(symbolReference.absoluteString.singleQuoted).
+                    Non-symbols should already have been filtered out in `PathHierarchyBasedLinkResolver.traverseOverloadedSymbols(_:)`.
+                    """)
+                }
+                guard symbolReference.sourceLanguage == .swift else {
```
[the remark this answers is not in the document]
Comment (QuietMisdreavus) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2422]: The current behavior, based on a quick test with this code, tells me the following:

- The `@objc` method correctly appears on the Objective-C view of `MyClass`.
- Selecting the `-> Int` version of the method in Swift enables the language picker on that page, as well as showing the overloads listing.
- Selecting the `-> String` version of the method disables the language picker, because there is no corresponding Objective-C method to switch to.

So i think it behaves correctly, from what i can tell. When the automatic curation tweaks are added so that only one page is linked in the curation hierarchy, we'll need to make sure that the individual overload selections still allow you to switch to the appropriate language view, but that's later work.

**Comment by QuietMisdreavus** [comment]
@swift-ci Please test
